### PR TITLE
Northwest Class 4 - Terry Calderbank - JS Core 1 - Week 2

### PR DESCRIPTION
**Pull request by Texx45**
**Volunteers: Are you marking this coursework?** _You can find a guide on how to mark this coursework in `HOW_TO_MARK.md` in the root of this repository_

# Your Details

- Your Name:
- Your City:
- Your Slack Name:

# Homework Details

- Module:
- Week:

# Notes

- What did you find easy?

- What did you find hard?

- What do you still not understand?

- Any other notes?

-----
[View rendered .github/pull_request_template.md](https://github.com/Texx45/JavaScript-Core-1-Coursework-Week2/blob/main/.github/pull_request_template.md)
[View rendered CODE_STYLE.md](https://github.com/Texx45/JavaScript-Core-1-Coursework-Week2/blob/main/CODE_STYLE.md)
[View rendered GRADING.md](https://github.com/Texx45/JavaScript-Core-1-Coursework-Week2/blob/main/GRADING.md)
[View rendered HOW-TO-GET-HELP.md](https://github.com/Texx45/JavaScript-Core-1-Coursework-Week2/blob/main/HOW-TO-GET-HELP.md)
[View rendered HOW-TO-SUBMIT.md](https://github.com/Texx45/JavaScript-Core-1-Coursework-Week2/blob/main/HOW-TO-SUBMIT.md)
[View rendered HOW_TO_MARK.md](https://github.com/Texx45/JavaScript-Core-1-Coursework-Week2/blob/main/HOW_TO_MARK.md)
[View rendered README.md](https://github.com/Texx45/JavaScript-Core-1-Coursework-Week2/blob/main/README.md)
[View rendered exercises/A-expressions/README.md](https://github.com/Texx45/JavaScript-Core-1-Coursework-Week2/blob/main/exercises/A-expressions/README.md)
[View rendered exercises/B-boolean-literals/README.md](https://github.com/Texx45/JavaScript-Core-1-Coursework-Week2/blob/main/exercises/B-boolean-literals/README.md)
[View rendered exercises/C-comparison-operators/README.md](https://github.com/Texx45/JavaScript-Core-1-Coursework-Week2/blob/main/exercises/C-comparison-operators/README.md)
[View rendered exercises/D-logical-operators/README.md](https://github.com/Texx45/JavaScript-Core-1-Coursework-Week2/blob/main/exercises/D-logical-operators/README.md)
[View rendered exercises/E-conditionals/README.md](https://github.com/Texx45/JavaScript-Core-1-Coursework-Week2/blob/main/exercises/E-conditionals/README.md)
[View rendered exercises/F-predicates/README.md](https://github.com/Texx45/JavaScript-Core-1-Coursework-Week2/blob/main/exercises/F-predicates/README.md)
[View rendered exercises/G-conditionals-2/README.md](https://github.com/Texx45/JavaScript-Core-1-Coursework-Week2/blob/main/exercises/G-conditionals-2/README.md)
[View rendered exercises/H-array-literals/README.md](https://github.com/Texx45/JavaScript-Core-1-Coursework-Week2/blob/main/exercises/H-array-literals/README.md)

### CODE REVIEW

File: exercises/G-conditionals-2/exercise-2.js [hunk 1]
```diff
@@ -1,29 +1,31 @@
...
+  Conditionals
+  ---------------------------------
+  Write a function that checks if a student has passed
+  - if the grade is less than 50 then return "failed"
+  - if 50 or higher then return "passed"
+
+*/
+
+function studentPassed(grade) {
+  return grade >= 50 ? `passed` : `failed`;
```
Comment (dharmaguadeloupe) [inline exercises/G-conditionals-2/exercise-2.js:11]: Nice use of ternary statements

File: mandatory/2-function-creation.js [hunk 2]
```diff
@@ -35,8 +36,11 @@ Write a function that:
 - makes all strings all lowercase
 */
 function tidyUpStrings(arrayOfStrings) {
-  const editedString = arrayOfStrings;
-  return editedString.trim().replace("/", "");
+  const stringArray = [];
+  for (let i = 0; i < arrayOfStrings.length; i++) {
+    stringArray.push(arrayOfStrings[i].trim().replace("/", "").toLowerCase());
```
[the remark this answers is not in the document]
Comment (xtremelibasic) [reply mandatory/2-function-creation.js:41]: I absolutely love your solution

**Review by xtremelibasic — COMMENTED** [review]
Great work Terr, Your code was very concise throughout.

File: mandatory/2-function-creation.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (Texx45) [reply mandatory/2-function-creation.js:41]: Thank you all much appreciated!

File: exercises/G-conditionals-2/exercise-2.js (hunk 1 above)
Comment (Texx45) [reply exercises/G-conditionals-2/exercise-2.js:11]: Thanks Dharma!

**Comment by SteveLeicester** [comment]
Nice tidy code.  Agree with Musa

`stringArray.push(arrayOfStrings[i].trim().replace("/", "").toLowerCase());`

is very nice.   It's method chaining and sometimes called fluent coding style.  

Found one or two things to watch out for.

Looking at G-conditionals-2 execise-3 

Couple of errors. Looks like you have used mark as the parameter to the function then used grade in function logic.  So grade is undefined.

`function calculateGrade(mark) {`
`  if (grade >= 80) {`
`    return "A";`
`  } else if (grade >= 60) {`
`    return "B";`
`  } else if (grade >= 50) {`
`    return "C";`
`  } else {`
`    return "F";`
`  }`
`}`


Also if the mark is greater than 60 then the grade is B. Your code will generously give the student with 60 a grade B when they should have got a C.  It's a good illustration - turning natural language into hard logic has a long long history of being a source of human error.  And software development has a long history of finding ways to catch errors.

It's a great illustration of why using test driven development (TDD) is so popular these days.  You would end up with a bunch of tests - looking at edge cases and failing in this case.

Beyond the exercises - all your tests pass. So that's what counts. But also code is for others to read, understand and change so it is important to keep things as simple as possible and try to avoid confusing humans. So your mood function works but there is no need to copy the parameter into a local variable  (often there may be good reason to do this). So here just use it directly. Like so



`function mood(isHappy) {`
`  if (isHappy) {`
`    return 'I am happy';`
`  } else {`
`    return 'I am not happy';`
`  }`
`}`

The greaterThan10  function is logically perfect. But choosing names for variables is to help humans. The **is** in isBigEnough suggest that this is a Boolean variable (takes True or False).  But it is given a number (from the parameter  num).  So how about

` let isBigEnough = num > 10; /* True or False */`
` if (isBigEnough) {          /* Test the booleam */`
`    . . . . .`
`    . . . . .`
`    etc	`


Don't think there is anything else - great to see you did the extras too.  Basically a fine piece of work again - Thanks Terry.

**Comment by Texx45** [comment]
Thanks for the review Steve, I was having a problem running jest in my terminal so I had to write the code outside of VSC and run the tests manually. Anthony helped me sort it out and I transferred it all back so I could run the tests.  I was aware of the grade/mark naming error as it errored for me and I thought I had changed it.

I will take on board the other feedback and I appreciate the advice. Thanks again Steve!
